### PR TITLE
[#819 ] feat(operator): Tez ApplicationMaster supporting RemoteShuffle

### PR DESCRIPTION
**Pull request by baitian77**
### What changes were proposed in this pull request?
1. RssDAGAppMaster will start rpc server and process getShuffleAssignments requests(assign uniffle works)   from map/reduce task to shuffle write/read
2. Keep those uniffle works heatbeat with coordinator in application lifecycle

### Why are the changes needed?

Fix: #819 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?
UT

### CODE REVIEW

File: client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+import static org.apache.tez.common.TezCommonUtils.TEZ_SYSTEM_SUB_DIR;
+
+public class RssDAGAppMaster extends DAGAppMaster {
+  private static final Logger LOG = LoggerFactory.getLogger(RssDAGAppMaster.class);
+  private ShuffleWriteClient shuffleWriteClient;
+  private TezRemoteShuffleManager tezRemoteShuffleManager;
+  private static final String rssConfFileLocalResourceName = "rss_conf.xml";
+  private DAGProtos.PlanLocalResource rssConfFileLocalResource;
+  final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
```
[the remark this answers is not in the document]
Comment (baitian77) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:78]: done

File: client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            appMaster.getTezRemoteShuffleManager().address.getHostName());
+    extraConf.setInt(RssTezConfig.RSS_AM_SHUFFLE_MANAGER_PORT,
+            appMaster.getTezRemoteShuffleManager().address.getPort());
+    writeExtraConf(appMaster, conf, extraConf, strAppId);
+
+    mayCloseTezSlowStart(conf);
+  }
+
+  @Override
+  public String submitDAGToAppMaster(DAGProtos.DAGPlan dagPlan, Map<String, LocalResource> additionalResources)
```
[the remark this answers is not in the document]
Comment (baitian77) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:161]: done

File: client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  }
+
+  /**
+   * Init and Start Rss Client
+   * @param appMaster
+   * @param conf
+   * @param applicationAttemptId
+   * @throws Exception
+   */
+  public static void initAndStartRSSClient(final RssDAGAppMaster appMaster, Configuration conf,
```
[the remark this answers is not in the document]
Comment (baitian77) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:113]: No need,  in tez DAGAppMaster alreay exits hadoopShim Object.   https://github.com/apache/tez/blob/master/tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java#L429

[the remark this answers is not in the document]
Comment (jerqi) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:113]: You don't get my point. You can read the comment
```
UpdateNodes only have one use for MRAppMaster, MRAppMaster use the updateNodes to find which
nodes are bad nodes. So we clear them, MRAppMaster will not recompute the map tasks.
```

[the remark this answers is not in the document]
Comment (baitian77) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:113]: Will be optimized in future PR.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @baitian77

**Comment by jerqi** [comment]
Congrats! 600th commits.

**Pull request merged**